### ISSUE TITLE
Visibility of automatic app upgrades in GitOps

### ISSUE THREAD

**Issue by gianfranco-l** [giantswarm/roadmap#2211]
In a recent feedback session with one of our customer, the following idea was discussed, and it's worth to analyze

> with Customer move to gitops, we will help them set up a system to automate the opening of PRs to upgrade apps (similar to something other customers already have). Ideally, the PRs should include information on what changes between the two versions, and it would be even better if we could check for possible problems in the git repo beforehand. E.g. nginx deprecates an old API and we check whether there are references to that API in the git repo, something like this. Of course it's a long term goal and what we talked about is all theoretical - in general we should enable customers to eventually "just approve a PR" for upgrading apps. The responsibility of following their upgrade schedule etc. is still on the customer. Same holds for configuration: customer owns it. Our goal should be to make customer just "consume" upgrades (click the button) without too much need to worry. Again - long-term goal to iterate on and refine.

**Comment by gianfranco-l** [comment]
also related to [this](https://github.com/giantswarm/giantswarm/issues/24072#issuecomment-1481568288), with @piontec we talked about having a lack of alignment on customers expectations + education on Flux capabilities. What are your inputs/ideas?
An idea that came into the conversation was to create an FAQ about use cases covered by Flux, with further readings linking to Flux docs.

**Comment by piontec** [comment]
@mproffitt I think Gianfranco forgot to mention you here - what do you think about the expectations of our customers and how we as a company manage them? Today we had the discussion about monitoring AWS resoruces for our custoemrs.

**Comment by gianfranco-l** [comment]
as team we agreed to channel all the inputs in the larger conversation on [standardize gitops offering](https://github.com/giantswarm/giantswarm/issues/26269) so closing in favor of that ticket

**Comment by LolloneS** [comment]
First of all: thank you!

For my understanding: is there some idea on what we think customers will be able to do in the future? For instance: do we think it makes sense for them to configure [this](https://fluxcd.io/flux/guides/image-update/)? Are we planning on some other mechanism to help customers keep their configs up-to-date? I know there are no details yet and it's all still WIP so any (even generic) indication is more than appreciated

**Comment by LolloneS** [comment]
I believe we are on the same page. Each customer has their specific implementation and of course we can't provide a one-size-fits-all, nor can we take responsibility of failures independent of our control. 

In other words: 
>it is partly about the question of who will manage the outcome of failure, what's likely to be the expectation there

it's the same as it is now: customers need to manage their repo. If something breaks, we usually let them know so they can fix it in their repo. In case of emergency, we can suspend their kustomizations if that is needed to fix issues. If a customer does not have the capacity to properly set up a GitOps repo, that's an onboarding package we can talk about that's separate from the usual managed service we provide, because that's basically consulting. I know we have had multiple experiences and it won't always be this easy, but the fundamental point I want to convey is: helping a customer enable automatic app upgrades does not make us "more responsible" for their Git repos than we are now.

What I think we should define is how we can help them get started with the automatic app upgrades topic, where needed. Flux has some capabilities in this direction and I think we can give our opinionated suggestion to customers. But the repo is theirs and remains theirs, so they need to take care of configuration, problems, etc.

As I mentioned on Slack, stuff such as `possibly the output of a check telling them if there are references to breaking changes in their git repo` is a long-term wish that might become reality in 1 year with some new Flux feature, who knows. It's nothing to be expected now, it's the dream scenario. Right now IMHO the first iteration we should define is: how do we think customers should configure automatic app upgrades? We currently have [something](https://github.com/giantswarm/gitops-template/blob/main/docs/apps/automatic_updates_appcr.md) in our `gitops-template` repo and we might just say that's how we suggest doing it, or rework it, depending on the current status of things. But that's about it. In addition to this, as you mention we provide tooling (and upstream does as well) to ease operations on both our and our customers' sides. 

Long story short: I'd love to take this step-by-step. The first thing IMHO is defining whether our vision on automatic app upgrades via Flux has changed with respect to what we have in `gitops-template`, possibly updating it, etc. Then we can think about next steps, cooler tools, and so on. But I don't think that showing customers how we think automatic app upgrades should be done and configuring a sample one with them will make us "own" their GitOps repo more than we do now.

I hope I explained myself decently, I'm here for any clarification/discussion :)

**Comment by piontec** [comment]
OK, I think we all agree, that we can only make small steps. And that customers are expected to make feature requests. Let's just make sure that we can meet somewhere on the way :) Currently, we're offering what is in the gitops-examples, with the validation based on `flux-build`. This can be easily incorporated into a git management system, like github, and run on PRs, but that's already in customer's responsibility area to configure. Examples of how we use that are in gitops-template (but we have to check if it's in-sync with some changes that went to management-clusters-fleet, as something might need backporting).

**Comment by LolloneS** [comment]
Thanks!

To sum up, we currently have:

- Documentation on [GitOps tooling](https://docs.giantswarm.io/advanced/gitops/tools/)
- The [gitops-template](https://github.com/giantswarm/gitops-template) repository (which may need some polishing)

In both of them, information on how to set up automatic App upgrades can be found. If customers have specific needs, we should open roadmap issues for specific feature requests.

Any objections to me closing this issue?

**Comment by mproffitt** [comment]
No objections from me. FTR the gitops tooling doc doesn't cover automatic upgrades but is about tools to support validating and testing manifests.

The main information is in gitops-template and we may want some clarity around this.

I would say close this and we'll raise a new issue for polishing the template as there may be other items we want to address as part of that work.
